libzecale/circuits/pairing/bls12_377_pairing.hpp

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef ZECALE_CIRCUITS_PAIRING_BLS12_377_PAIRING_HPP__
+#define ZECALE_CIRCUITS_PAIRING_BLS12_377_PAIRING_HPP__
+
+#include "libzecale/circuits/fields/fp12_2over3over2_gadgets.hpp"
+
+#include <libff/algebra/curves/bls12_377/bls12_377_pp.hpp>
+#include <libsnark/gadgetlib1/gadgets/fields/fp2_gadgets.hpp>
+#include <libsnark/gadgetlib1/gadgets/pairing/pairing_params.hpp>
+
+namespace libzecale
+{
+
+/// Holds an element of G2 in homogeneous projective form. Used for
+/// intermediate values of R in the miller loop.
+template<typename ppT> class bls12_377_G2_proj
+{
+public:
+    libsnark::Fqe_variable<ppT> X;
+    libsnark::Fqe_variable<ppT> Y;
+    libsnark::Fqe_variable<ppT> Z;
+
+    bls12_377_G2_proj(
+        libsnark::protoboard<libff::Fr<ppT>> &pb,
+        const std::string &annotation_prefix);
+
+    bls12_377_G2_proj(
+        const libsnark::Fqe_variable<ppT> &X_var,
+        const libsnark::Fqe_variable<ppT> &Y_var,
+        const libsnark::Fqe_variable<ppT> &Z_var);
+
+    void evaluate() const;
+
+    void generate_r1cs_witness(const libff::bls12_377_G2 &element);
+};
+
+/// Not a gadget - holds the variables for the Fq2 coefficients of the tangent
+/// line at some R, used during the doubling step.
+template<typename ppT> class bls12_377_ate_ell_coeffs
+{
+public:
+    const libsnark::Fqe_variable<ppT> ell_0;
+    const libsnark::Fqe_variable<ppT> ell_vw;
+    const libsnark::Fqe_variable<ppT> ell_vv;
+
+    bls12_377_ate_ell_coeffs(
+        const libsnark::Fqe_variable<ppT> &ell_0,
+        const libsnark::Fqe_variable<ppT> &ell_vw,
+        const libsnark::Fqe_variable<ppT> &ell_vv);
+
+    void evaluate() const;
+};
+
+/// Gadget that relates some "current" bls12_377_G2_proj value in_R with the
+/// result of the doubling step, that is some bls12_377_G2_proj out_R and the
+/// bls12_377_ate_ell_coeffs holding the coefficients of the tangent at in_R.
+/// Note that the output variables are allocated by this gadget.
+template<typename ppT>
+class bls12_377_ate_dbl_gadget : libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    typedef libff::Fq<libsnark::other_curve<ppT>> FqT;
+    typedef libff::Fqe<libsnark::other_curve<ppT>> FqeT;
+
+    bls12_377_G2_proj<ppT> in_R;
+    bls12_377_G2_proj<ppT> out_R;
+
+    // TODO: Many of these intermediate Fqe_variables are only for clarity and
+    // replicate the references held by other gadgets (e.g. `A` refers to the
+    // same variable as `check_A.result`. Do an optimization pass and remove
+    // some of the redundancy.
```

Comment:
@dtebbs I assume this comment refers to what we discussed off-line this am, is that right?

Reply:
Yes.  I think we covered this in a discussion.  It should be addressed in a later PR.